src/algebra/monoid_algebra.lean

```diff
@@ -373,6 +377,85 @@ end
 
 end misc_theorems
 
+/-! #### Non-unital, non-associative algebra structure -/
+section non_unital_non_assoc_algebra
+
+variables {R : Type*} (k) [semiring R] [semiring k] [distrib_mul_action R k] [has_mul G]
+
+instance is_scalar_tower_self [is_scalar_tower R k k] :
+  is_scalar_tower R (monoid_algebra k G) (monoid_algebra k G) :=
+⟨λ t a b,
+begin
+  ext m,
+  simp only [mul_apply, finsupp.smul_sum, smul_ite, smul_mul_assoc, sum_smul_index', zero_mul,
+     if_t_t, implies_true_iff, eq_self_iff_true, sum_zero, coe_smul, smul_eq_mul, pi.smul_apply,
+     smul_zero],
+end⟩
+
+/-- Note that if `k` is a `comm_semiring` then we have `smul_comm_class k k k` and so we can take
+`R = k` in the below. In other words, if the coefficients are commutative amongst themselves, they
+also commute with the algebra multiplication. -/
+instance smul_comm_class_self [smul_comm_class R k k] :
+  smul_comm_class R (monoid_algebra k G) (monoid_algebra k G) :=
+⟨λ t a b,
+begin
+  ext m,
+  simp only [mul_apply, finsupp.sum, finset.smul_sum, smul_ite, mul_smul_comm, sum_smul_index',
+    implies_true_iff, eq_self_iff_true, coe_smul, ite_eq_right_iff, smul_eq_mul, pi.smul_apply,
+    mul_zero, smul_zero],
+end⟩
+
+instance smul_comm_class_symm_self [smul_comm_class k R k] :
+  smul_comm_class (monoid_algebra k G) R (monoid_algebra k G) :=
+⟨λ t a b, by { haveI := smul_comm_class.symm k R k, rw ← smul_comm, } ⟩
+
+variables {A : Type u₃} [non_unital_non_assoc_semiring A] [module k A]
+
+/-- A non_unital `k`-algebra homomorphism from `monoid_algebra k G` is uniquely defined by its
+values on the functions `single a 1`. -/
+lemma non_unital_alg_hom_ext ⦃φ₁ φ₂ : non_unital_alg_hom k (monoid_algebra k G) A⦄
+  (h : ∀ x, φ₁ (single x 1) = φ₂ (single x 1)) : φ₁ = φ₂ :=
```

Comment:
Does this generalize for `distrib_mul_action`? I haven't checked because gitpod is broken.

Reply:
Yes, I believe it is still true if we replace `[module k A]` with `[distrib_mul_action k A]`.

It looks like it might be quite a bit of work to generalise it since _I think_ almost all of the `finsupp` library demands a `module`, `linear_map` even when `distrib_mul_action`, `distrib_mul_action_hom` might suffice.

It's worth trying though. I'll have a go now.

Reply:
OK I have this working locally and it was not too much work. I'll tidy it up and push this evening.

Reply:
OK I just pushed https://github.com/leanprover-community/mathlib/pull/7932/commits/42ac4237f7ba07813b4e758bb231e1f0d6456843 which adds this generalisation.